Add wait_for_worker param to TCPStore and fix port in use flaky test failures

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52888 Add wait_for_worker param to TCPStore and fix port in use flaky test failures**

Changes:
1) Add optional wait_for_workers param in TCPStore Python constructor to match C++ constructor.
2) Fix intermittent "Address already in use" errors in tests due to race condition (#52839, #52770)

Differential Revision: [D26678707](https://our.internmc.facebook.com/intern/diff/D26678707)

H-Huang committed:
Add wait_for_worker param to TCPStore and fix port in use flaky test failures

[ghstack-poisoned]

cbalioglu commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1466,14 +1466,12 @@ def download_file(url, binary=True):
-    sockname = sock.getsockname()
-    sock.close()
-    return sockname[1]
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
```
Might not be relevant for this particular PR, but there is still a race condition here. Once the socket gets closed at the end of this call, there is no guarantee that the returned port number will be available for later use. Another process can grab it in-between.

Another solution could be to reserve a pool of port numbers in advance and hand them out as requested.

H-Huang replied:
Yes good catch. This function is called with a retry (seen in create_tcp_store()) to handle this race condition, so the main fix in this PR was to move the existing tests under that retry which will fix the flakiness.

cbalioglu commented on test/distributed/test_c10d.py:
```diff
@@ -355,54 +353,42 @@ def test_compare_set(self):
+        error_message = ""
+        while not messages.empty():
+            error_message += messages.get() + "\n"
+        if error_message:
```
A process might also fail abruptly without a clean path (e.g. OOM, SEGV). I suggest checking the exit codes instead of the length of the `error_message` here.

H-Huang replied:
Thanks! I updated the check

cbalioglu commented on test/distributed/test_c10d.py:
```diff
@@ -355,54 +353,42 @@ def test_compare_set(self):
+    def _create_client(self, addr, port, world_size, messages):
         try:
             client_store = dist.TCPStore(addr, port, world_size, timeout=timedelta(seconds=10))
             self.assertEqual(b"value", client_store.get("key"))
```
I don't see any assertions for `set` and `compare_set`. Do we call them to verify that we don't get any `RuntimeError`? If that is the case, I recommend adding a quick note here.

cbalioglu reviewed (commented):
Mostly looks good to me.

H-Huang committed:
Update on "Add wait_for_worker param to TCPStore and fix port in use flaky test failures"


Changes:
1) Add optional wait_for_workers param in TCPStore Python constructor to match C++ constructor.
2) Fix intermittent "Address already in use" errors in tests due to race condition (#52839, #52770)

Differential Revision: [D26678707](https://our.internmc.facebook.com/intern/diff/D26678707)

[ghstack-poisoned]

cbalioglu commented on test/distributed/test_c10d.py:
```diff
@@ -355,54 +353,43 @@ def test_compare_set(self):
+        error_message = ""
+        while not messages.empty():
+            error_message += messages.get() + "\n"
+        if any(map(lambda p: p.exitcode != 0, processes)):
```
nitpick: You can use a generator expression instead of `map` here.

H-Huang committed:
Update on "Add wait_for_worker param to TCPStore and fix port in use flaky test failures"


Changes:
1) Add optional wait_for_workers param in TCPStore Python constructor to match C++ constructor.
2) Fix intermittent "Address already in use" errors in tests due to race condition (#52839, #52770)

Differential Revision: [D26678707](https://our.internmc.facebook.com/intern/diff/D26678707)

[ghstack-poisoned]

H-Huang committed:
Update on "Add wait_for_worker param to TCPStore and fix port in use flaky test failures"


Changes:
1) Add optional wait_for_workers param in TCPStore Python constructor to match C++ constructor.
2) Fix intermittent "Address already in use" errors in tests due to race condition (#52839, #52770)

Differential Revision: [D26678707](https://our.internmc.facebook.com/intern/diff/D26678707)

[ghstack-poisoned]

H-Huang committed:
Update on "Add wait_for_worker param to TCPStore and fix port in use flaky test failures"


Changes:
1) Add optional wait_for_workers param in TCPStore Python constructor to match C++ constructor.
2) Fix intermittent "Address already in use" errors in tests due to race condition (#52839, #52770)

Differential Revision: [D26678707](https://our.internmc.facebook.com/intern/diff/D26678707)

[ghstack-poisoned]
